docs/user-guide/configuring.md

```diff
@@ -713,6 +713,61 @@ module.exports = {
 }
 ```
 
+## Configuration Based on Glob Patterns
+
+Sometimes a more fine-controlled configuration is necessary, for example if the configuration for files within the same directory has to be different. Therefore you can provide configurations under the `overrides` key that will only apply to files that match specific glob patterns, using the same format you would pass on the command line (e.g., `app/**/*.test.js`).
+
+### How it works
+
+* Glob pattern overrides can only be configured within `.eslintrc` files.
+* The patterns are applied against the file path relative to the directory of the `.eslintrc` file. For example, if your `.eslintrc` file has the path `/Users/john/workspace/any-project/.eslintrc` and the file you want to lint has the path `/Users/john/workspace/any-project/lib/util.js`, then the pattern provided in `.eslintrc` will be executed against the relative path `lib/util.js`.
+* Glob pattern overrides have higher precedence than the regular configuration in the same `.eslintrc` file.
+* A glob specific configuration works almost the same as the regular configuration in your `.eslintrc`, except that you can't define nested glob based configurations or use `extends`.
+* Multiple glob patterns can be provided within a single override block. The associated configuration will be applied to the intersection of the sets of files matched by each of the patterns. In other words, a file must match all the supplied patterns for the configuration to apply.
```

Comment:
I don't wanna bikeshed this too much because I really want to see this merged!  But the intersection behavior would confuse me based on my experience with other configuration systems; I would expect it to be a union behavior.  I don't think the `{foo,bar}` glob syntax is very well known.  Maybe if it's well-documented here, that's enough?

Here's another option, though.  What if `files` as an array had union semantics, but it could also optionally be an object with `includes` and `excludes` keys.  An array would be equivalent to `files: { includes: [...] }`.

With this, `["foo.js", "bar.js"]` would work as @not-an-aardvark expects, and @smably's use case could be handled as: `files: { includes: ["foo/**/*.js"], excludes: ["foo/**/*.test.js"] }`.  `excludes` could also be `except` to better communicate that it comes after the inclusion.